### ISSUE TITLE
return A() can't be assigned to Self@A

### ISSUE THREAD

**Issue by digulla** [microsoft/pyright#10787]
**Describe the bug**

This code fails the type check with the error "The type "A" can't be assigned the to type "Self@A".

**Code or Screenshots**

```python
from typing import Self


class A:
    def error1(self) -> Self:
        """A can't be assigned to Self@A"""
        return A()  # <-- error

    def error2(self) -> A:  # <-- error
        """A isn't known at this time, so I can't use it in the method definition."""
        return A()

    def works(self) -> "A":
        """This works."""
        return A()
```

I also can't use `def bar(self) -> A:` since the type isn't complete at that time. With this version of pyright, I'm forced to `def bar(self) -> "A":`

What's the correct solution in this case?

**VS Code extension or command-line**
Python 3.12
Identifier ms-python.vscode-pylance
Version 2025.7.1

**Comment by erictraut** [comment]
Pyright's behavior here is correct. The value of `A()` is not compatible with a return type of `Self`. For more details about the `Self` type, refer to [this part of the Python typing spec](https://typing.python.org/en/latest/spec/generics.html#self).

If your intent is to always return an instance of `Self`, you'd need to change the implementation of `error1` to something like this:

```python
    def error1(self) -> Self:
        return type(self)()
```

or

```python
    def error1(self) -> Self:
        return self.__class__()
```

If your intent is to return `A()` in all cases — even for subclasses of `A`, then `-> A` is the correct return type annotation. However, as you noted, you must use quotes because this is a forward reference. Alternatively, you can add `from __future__ import annotations` at the top of your code file. In Python 3.14, this is no longer necessary because all type annotation evaluation is deferred in 3.14.
